### PR TITLE
add implementation of Generating Action Sequences for Robot Task Plan using LLM

### PR DESCRIPTION
**Pull request by Undertone0809**
Add implementation of Generating Action Sequences for Robot Task Plan using LLM.

Some screenshots.

image

image

### CODE REVIEW

**Comment by JoeyHwong-gk** [comment]
Thank you for your pull request and the contribution to Sig Robotics.
There are a few points we'd like to address:

1. **Use of Third-Party or Unrelated Dependencies:**
   We noticed that your PR includes a significant number of third-party or unrelated components that are not directly related to the project's objectives. To maintain project cleanliness and ensure only necessary components are used, we recommend replacing these with container instances wherever possible. This will help streamline the project's dependencies and make it easier to manage.

2. **OSPP 2023 Project Guidelines:**
   As this is an OSPP 2023 project, it's important to ensure that your PR link and the content submitted align with the project proposal. We encourage you to maintain consistency between your PR and the [proposal](https://github.com/kubeedge/community/pull/146) to ensure that your work aligns with the project's goals and objectives.

3. **Unfinished Work:**
   If there are parts of your PR that are incomplete or still under development, we recommend marking them as drafts. This will help reviewers and contributors understand the status of your work and avoid any confusion during the review process.

4. **Commit Signing and KubeEdge Header:**
   Please ensure that the commits you wish to merge are signed (DCO requires). Additionally, for any personal code contributions, make sure to use KubeEdge's header as per project guidelines.

Feel free to make the necessary adjustments and updates to your PR, and we will review it accordingly. If you have any questions or need further assistance, please don't hesitate to contact your mentor (@wjmstf1) . 
Your efforts are valued, and we look forward to collaborating with you on this project.

**Comment by Undertone0809** [comment]
Received, I will update the relevant content.

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+FROM ros:ROS_DISTRO
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/Dockerfile:1]: Maybe you mean "FROM ros:humble", you need to clarify the version here

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+## Setup
+
+**Clone the community/sig-robotics respository**
+
+```shell
+git clone https://github.com/kubeedge/community
+```
+
+```shell
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/README.md:18]: compile after entering the project

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/setup.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        ("share/ament_index/resource_index/packages", ["resource/" + package_name]),
+        ("share/" + package_name, ["package.xml"]),
+        (os.path.join("share", package_name, "launch"), glob("launch/*.launch.py")),
+        (os.path.join("share", package_name, "urdf"), glob("urdf/**")),
+        (os.path.join("share", package_name, "world"), glob("world/**")),
+    ],
+    install_requires=["setuptools", "promptulate"],
+    zip_safe=True,
+    maintainer="zeeland",
+    maintainer_email="zeeland@foxmail.com",
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/setup.py:21]: Respecting others' contributions is a fundamental principle of open source. Please avoid invasive changes. You can run your node in a separate terminal while keeping the FishROS node running in another terminal. This way, both nodes can operate independently without interference.

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/bot_description/core.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        return sensor.get_data()
+
+    def get_sensor(self, sensor_name: str) -> Sensor:
+        return next(
+            (sensor for sensor in self.sensors if sensor.name == sensor_name),
+            None,
+        )
+
+    def show_all_data(self):
+        data = ""
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/bot_description/core.py:60]: more Pythonic: `data = "".join([str(sensor.get_data()) for sensor in self.sensors])`

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/bot_description/core.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    def get_operator(self, action_name: str) -> Operator:
+        return next(
+            (operator for operator in self.operators if operator.name == action_name),
+            None,
+        )
+
+
+class RobotObserver:
+    def __init__(self, sensors: List[Sensor]) -> None:
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/src/bot_description/bot_description/core.py:46]: The sensor data does not have semantic information, which may cause a gap for llm.

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/README.md [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+rosdep install --from-paths src -y
+colcon build
+```
+
+**Gazebo in gazebo**
```
Comment (JoeyHwong-gk) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/README.md:23]: update here to make it more clearer

**Comment by JoeyHwong-gk** [comment]
Thank you for your contribution. This PR represents significant progress in enabling robot movement based on natural language instructions. I appreciate your efforts. Here are a few suggestions for further improvements:

1. It appears that an unnecessary commit [02bbd19f163ff5beef2b778b025d9aff108f491d] was introduced, possibly due to a sync during your development process. I recommend removing it and consolidating your commits through rebase for clarity.

2. As part of our community guidelines, please ensure that all executable scripts include the KubeEdge license.

3. Kindly move your proposal into this project repository and close the previous PR related to it.

4. Although you have declared dependencies in setup.py, it's advisable to maintain these dependencies in requirements.txt for better management.

5. It seems you are launching Gazebo directly from the code. Consider launching Gazebo in a separate container and connecting to it through the network. This is just a suggestion for optimization.

6. This work seems promising but requires further development to achieve the goals outlined in your proposal. I look forward to seeing your continued efforts in this area.

**Comment by wjmstf1** [comment]
A more complex task will better show the benifits of LLM.

**Review by fisherxu — COMMENTED** [review]
/approve

File: sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/.gitmodules [hunk 7]
```diff
@@ -0,0 +1,10 @@
+[submodule "src/navigation"]
+	path = src/navigation
+	url = https://github.com/fishros/navigation2.git
```
Comment (fujitatomoya) [inline sig-robotics/examples/Generating-Action-Sequences-for-Robot-Task-Plans-using-LLM/.gitmodules:3]: (just out of curiosity) Why this is forked from mainline? can we pull with using specific hash or this includes some specific code or requirements?

**Review by fujitatomoya — COMMENTED** [review]
@Undertone0809 i know it is too late to review, but i got a minor question. thanks for sharing the concreate and interesting proposal.

**Pull request merged**